### PR TITLE
patch: add  --skip-not-applicable-patches to  autoresolve not applicable patches

### PR DESCRIPTION
**Pull request by mlandres**
(closes #514)
The first commits are fixes and enhancements to the patch. The last one introduces the feature.

If multiple problems are to be reported, zypper uses a 1st pass to quickly list all problems and a 2nd pass to ask to resolve each problem. 

We use the 1st pass to check for auto-applicable solutions and re-solve if solutions were applied. If problems still remain at the end, we ask.

The draft is basically ready. Remaining tasks:

- [x] Remember skipped patches and show them in the summary.
- [x] Upadte man page.

### CODE REVIEW

**Comment by mlandres** [comment]
Screen output will look like this:
```
Resolving package dependencies...
2 Problems:
Problem: 1: the to be installed patch:openSUSE-SLE-15.4-2023-4534-1.noarch conflicts 
with 'zypper-aptitude < 1.14.66-150400.3.35.1' provided by the installed 
zypper-aptitude-1.14.61-150400.3.24.1.noarch
 Autoresolve:                                        (--skip-not-applicable-patches)
 Solution 1: Following actions will be done:
  do not install patch:openSUSE-SLE-15.4-2023-4534-1.noarch
  do not install patch:openSUSE-SLE-15.4-2023-3973-1.noarch
Problem: 2: the to be installed patch:openSUSE-SLE-15.4-2023-3514-1.noarch conflicts 
with 'zypper-aptitude < 1.14.63-150400.3.29.1' provided by the installed 
zypper-aptitude-1.14.61-150400.3.24.1.noarch
 Autoresolve:                                        (--skip-not-applicable-patches)
 Solution 1: Following actions will be done:
  do not install patch:openSUSE-SLE-15.4-2023-3514-1.noarch
  do not install patch:openSUSE-SLE-15.4-2023-3973-1.noarch

Resolving dependencies...
```

File: src/solve-commit.cc [hunk 1]
```diff
@@ -213,49 +236,89 @@ static TriBool show_problem( Zypper & zypper, const ResolverProblem & prob, Prob
...
+      if ( mayAutoResolve ) {
+        unsigned s = 0;
+        for ( const auto & solPtr : probPtr->solutions() )
+        {
+          if ( policy.skipNotApplicablePatches() ) {
+            std::optional<std::set<PoolItem>> items { solPtr->getIfSkipsPatchesOnly() };
+            if ( items ) {
+              pendingSolutions.push_back( solPtr );
+              policy.summaryHints.skippedPatchesSet().merge( *items );
+              if ( not printAbsatract ) { // was not printed before
```
Comment (bzeller) [inline src/solve-commit.cc:282]:
```suggestion
              if ( not printAbstract ) { // was not printed before
```

File: src/solve-commit.cc [hunk 2]
```diff
@@ -213,49 +236,89 @@ static TriBool show_problem( Zypper & zypper, const ResolverProblem & prob, Prob
...
+                dumpProblem( zypper.out().info(), p, *probPtr );
+              } else {
+                // Once we auto-resolved at least one problem, we will solve again before
+                // asking the user anything. So we stop printing any summary and list only
+                // auto-resolved problems.
+                printAbsatract = false;
```
Comment (bzeller) [inline src/solve-commit.cc:288]:
```suggestion
                printAbstract = false;
```

**Review by bzeller — CHANGES_REQUESTED** [review]

File: src/Summary.cc [hunk 3]
```diff
@@ -1404,6 +1498,15 @@ void Summary::writeRebootNeeded( std::ostream & out )
   checkAndPrintRebootNeeded( ResKind::package );
 }
 
+void Summary::writeAutoResolved( std::ostream & out )
```
[the remark this answers is not in the document]
Comment (mlandres) [reply src/Summary.cc:1501]: Well, ATM ist's just patches, but who knows. It's intended to be the hook for writing data from the summaryHints. Like the other writeSomething  methods it's not restricted to a specific Kind.

File: src/solve-commit.cc [hunk 4]
```diff
@@ -213,49 +236,89 @@ static TriBool show_problem( Zypper & zypper, const ResolverProblem & prob, Prob
...
+      if ( printAbsatract ) {
+        dumpProblem( zypper.out().info(), p, *probPtr );
+      }
+
+      if ( mayAutoResolve ) {
+        unsigned s = 0;
+        for ( const auto & solPtr : probPtr->solutions() )
+        {
+          if ( policy.skipNotApplicablePatches() ) {
+            std::optional<std::set<PoolItem>> items { solPtr->getIfSkipsPatchesOnly() };
```
[the remark this answers is not in the document]
Comment (mlandres) [reply src/solve-commit.cc:278]: And if you look down the context, one more line you find the not-const usage in `merge`:
```policy.summaryHints.skippedPatchesSet().merge( *items );``` 
which could actually use `std:move( *items )`

[the remark this answers is not in the document]
Comment (mlandres) [reply src/solve-commit.cc:278]: 281ff now:
```
              policy.summaryHints.skippedPatchesSet().merge( std::move(*items) );
              items.reset();  // moved out
```

**Review by bzeller — APPROVED** [review]

**Pull request merged**